### PR TITLE
Hoist statically bound allocations to the entry block of a function.

### PR DESCRIPTION
**Pull request by MaheshRavishankar**
All allocations on the LLVM-CPU backend are statically bound. But the way they are lowered today, it might end up as an allocation that is dynamic within some inner loop. Hoist such allocations to the entry block, and use a statically defined stack allocation instead.

### CODE REVIEW

**Review by benvanik — APPROVED** [review]

File: compiler/src/iree/compiler/Codegen/LLVMCPU/Passes.cpp [hunk 1]
```diff
@@ -90,11 +90,63 @@ struct LinalgCPUVectorLoweringPassOptions : LinalgVectorLoweringPassOptions {
...
+  {
+    OpBuilder::InsertionGuard g(builder);
+    builder.setInsertionPointToStart(&funcOp.getBody().front());
+    auto allocationType =
+        MemRefType::get(staticShape, memRefType.getElementType());
+    allocation = builder.create<memref::AllocaOp>(
+        loc, allocationType, builder.getI64IntegerAttr(alignment));
+  }
+
+  Value subviewOp = builder.create<memref::SubViewOp>(loc, allocation, offsets,
```
Comment (vmurali) [inline compiler/src/iree/compiler/Codegen/LLVMCPU/Passes.cpp:132]: Clarification: any subviews can be subsequently written into, right? If so, all this looks okay.

**Review by vmurali — COMMENTED** [review]
A unit test for both dynamic and static cases (inner memref.alloc by store into the allocated buffer) would help keep this code tested.

File: compiler/src/iree/compiler/Codegen/LLVMCPU/Passes.cpp (hunk 1 above)
Comment (MaheshRavishankar) [reply compiler/src/iree/compiler/Codegen/LLVMCPU/Passes.cpp:132]: yeah, you can do whatever with it. Its just an allocation.

**Comment by vmurali** [comment]
> A unit test for both dynamic and static cases (inner memref.alloc by store into the allocated buffer) would help keep this code tested.

Perhaps this is more complicated since you need to invoke bufferization? Feel free to ignore.

**Comment by MaheshRavishankar** [comment]
> > A unit test for both dynamic and static cases (inner memref.alloc by store into the allocated buffer) would help keep this code tested.
> 
> Perhaps this is more complicated since you need to invoke bufferization? Feel free to ignore.

Yeah it is... With https://github.com/iree-org/iree/pull/11938 it will become an error if we dont do this....

**Pull request merged**